tests/test_backoff.py

```diff
@@ -345,9 +345,32 @@ def test_backoff_many_conns(mock_ioloop_current):
     assert r.backoff_timer.get_interval() == 0
 
     for c in conns:
+        filtered_expected_args = []
+        i = 0
+        while i < len(c.expected_args):
+            j = i
+            while c.expected_args[i].startswith(b'RDY') and j + 1 < len(c.expected_args) \
+                    and c.expected_args[j + 1] == c.expected_args[i]:
+                j = j + 1
+                continue
+            rdy_found = False
+            for index in range(len(filtered_expected_args) - 1, -1, -1):
+                arg = filtered_expected_args[index]
+                if arg.startswith(b'RDY'):
+                    rdy_found = True
+                    if arg != c.expected_args[i]:
+                        if not (len(
+                                filtered_expected_args) == 1 and c.expected_args[i] == b'RDY 0\n'):
```

Comment:
This code to generate the expected output is ... too complicated. It is a too hard to read it and convince yourself that the test is basically right.

I can see how you got here. The assumption is that the library code is simple and correct, so the failing test needs to be fixed. But you went down the wrong path. I saw about 8 attempts pushed up over about 8 hours as you were working on this ... good evidence that this strategy is too complicated. There has to be a better way.

Maybe, instead of fixing-up `expected_args` at the end, the fixes should be while appending to `expected_args` throughout the test. But also, maybe the test setup is a bit "buggy" - this test does tricky mocking of the ioloop.

Reply:
Yes. It took me about 10 hours to fix the test. I will take a look at this to make it more readable.